KT-34617: `Collection` instances for `Progression`s

zhelenskiy opened the pull request:
`Collection` instances for `Char`, `Int`, `Long`, `UInt`, `ULong` `Progression`s.

*(Tests are also provided)*

Issue: [KT-34617](https://youtrack.jetbrains.com/issue/KT-34617)

Requires bootstrap advancing

ilya-g commented on .idea/codeStyles/Project.xml:
```diff
@@ -18,6 +18,9 @@
     </MarkdownNavigatorCodeStyleSettings>
+    <ScalaCodeStyleSettings>
+      <option name="MULTILINE_STRING_CLOSING_QUOTES_ON_NEW_LINE" value="true" />
+    </ScalaCodeStyleSettings>
```
Please remove changes of IDEA settings

zhelenskiy replied:
> Please remove changes of IDEA settings

Ok, I'll delete it after getting some response on the reply below.

ilya-g reviewed (changes requested):
- The first commit doesn't seem to be related to the proposed change, please remove it from your branch.
- Please split the branch clearly in two parts: first, the commits that don't require new bootstrap, then the commits that do. Without that it would be hard to run tests.

zhelenskiy commented:
* I'll delete the first commit somehow (but there was a bug introduced by someone else that broke compilation of hole project, and the first commit is cherry-picked from the main branch to fix it, so I don't know how to delete the commit and keep the project compilable, but recreating the fork)
* All the rest commits require bootstrap forwarding (the first is the main one, the rest are just corrections of some typos and so on: for example, when I was debugging I added an extra simple assertion to make it easier to catch and debug and forgot to remove it. Then I removed it in the some of the next commits/)

ilya-g commented:
You need to pull the recent commits from the master branch and rebase your master branch on top of these changes.
You can try `git pull -r` command for that. (Better to create a branch with your current changes for backup).

Next time it's better to create a feature branch in your fork, this way it would be easier to update the master branch later and rebase your feature branch on that.

zhelenskiy commented:
> You need to pull the recent commits from the master branch and rebase your master branch on top of these changes.
> You can try `git pull -r` command for that. (Better to create a branch with your current changes for backup).
> 
> Next time it's better to create a feature branch in your fork, this way it would be easier to update the master branch later and rebase your feature branch on it.

I have already coped with uniting my commits into the one.

zhelenskiy commented:
@ilya-g I have no extra files and commits for the PR (so it is ready, I suppose). I have fixed all you asked.

ilya-g commented:
Please extract test changes that depend on progression being a collection to a separate commit, this would greatly simplify the build that requires bootstrap advancing.

ilya-g commented on core/builtins/src/kotlin/Progressions.kt:
```diff
@@ -110,6 +122,18 @@ public open class IntProgression
+
+    override fun contains(@Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE") /* for the backward compatibility with old names */ value: Int): Boolean = when {
+        this.isEmpty() -> false
+        step > 0 && value >= first && value <= last -> (value - first) % step == 0
```
I wonder if `(value - first) % step` is overflow safe?
For example, `2147483649 % 3 == 0`, but the corresponding overflown value `-2147483647 % 3 == -1`

zhelenskiy replied:
I will replace it with `abs(value % step) == abs(first % step)` and will add some tests for overflow.

ilya-g replied:
I don't think it will do. Consider for example, a progression with `first = -1` and `step = 4`: the `value = 3` should be in it, but it has a different remainder than -1 has.

zhelenskiy replied:
As I haven't run the tests yet, I haven't found that. Thank you! (I still haven't run them, but I think that using `mod` from [ticket](https://youtrack.jetbrains.com/issue/KT-26234) would help).

zhelenskiy commented on generators/builtins/progressions.kt:
```diff
@@ -93,6 +96,18 @@ public open class $progression
+        get() = $sizeBody
+
+    override fun contains(@Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE") /* for the backward compatibility with old names */ value: $t): Boolean = when {
+        this.isEmpty() -> false
```
Actually, it is. I will remove it.

zhelenskiy commented on generators/builtins/progressions.kt:
```diff
@@ -93,6 +96,18 @@ public open class $progression
     override fun toString(): String = ${"if (step > 0) \"\$first..\$last step \$step\" else \"\$first downTo \$last step \${-step}\""}
 
+    override val size: Int
+        get() = $sizeBody
```
While implementing overflow handling for progressions I had steps `-1`, `1` as special cases for `size` (I will describe the cause after publishing the commit), so it is already implemented there, not in ranges.

The next question is about the `size` evaluation of `UInt` and `ULong` `Progressions`. Do they really need a special case `step = 1` for optimisation? Is dividing by `1.toUInt()` so expensive that we need to handle it as special case?

ilya-g replied:
Ok, the check for `step == 1` would be enough to cover the common case of primitive ranges inheriting progressions without overriding `size` there.

zhelenskiy replied:
Ok, then I will add the `step == 1` case to `UIntProgression` and `ULongProgression` even though I find the check useless.

zhelenskiy commented:
I have updated sources. I also generated prooves of overflow-safety of the `size` implementation (for `Long`s and `Int`s) in the comments. There are now two commits as you asked: the first is a bootstrap change, the second one is the tests change.

---

There were 2 reasons for overflow in `(last - first) / step + 1`:
1. `last - first`
  * The problem exists only for signed types as unsigned ones have subtraction result inside bounds of the type if the first operand is not less than the second one.
  * It is a bad idea to set `size` to Int.MAX_VALUE if `last - first` leads to overflow because dividing by step may decrease the real size and that would be less than `Int.MAX_VALUE`.
  * As described in the previous point, there would be 'real' overflow only if `step` is 1 or -1. Even `(Int.MAX_VALUE.toLong() - Int.MIN_VALUE) / 2` is equal to `Int.MAX_VALUE.toLong()`. 
  * So cases of `step` being 1 or -1 are cornercases and overflow check is trivial there.
  * If `|step| >= 2` I use that `(last - first) / step = (last / step * step + last % step - first / step * step - first % step) / step = last / step - first / step + (last % step - first % step) / step`.
  * If I can guarantee that `(last % step - first % step)` leads to no overflow, I prove that the whole formula has no overflow problems. (Proof is in the comments)
  * If I cannot guarantee that (if `|step|` is too big), I show (also in comments) that number of items in such progression is not more than 4, so it is easier and faster to check by iterating than to think up a formula to count them.
2. `+ 1` and converting `Long`, `UInt`, `ULong` to `Int`.
  * The check is trivial: if the number is *STRICTLY* less than `Int.MA_VALUE`, I increment it and convert to `Int`. Otherwise, the result is `Int.MAX_VALUE`.

---

The overflow problem with the `contains` operator was solved by comparing _(unsigned)_ `mod` operation.

ilya-g commented on libraries/stdlib/test/ranges/RangeTest.kt:
```diff
@@ -5,11 +5,198 @@
+        assertEquals(element in listRange, element in range, "$element in (expected $listRange) (found $range)")
+    }
+    for (element in bigRange) {
+        assertEquals(element in listRange, element in range)
```
What is the difference from the assertion in the loop above?

ilya-g replied:
ok, will remove

ilya-g commented on libraries/stdlib/test/ranges/RangeTest.kt:
```diff
@@ -5,11 +5,198 @@
+) {
+    for (start in limitRange)
+        for (finish in limitRange)
+            for (step in stepRange) {
```
This amount of iterations times out when running in JS. 
For example, in `longProgressionCollection` test, it's invoked with arguments that makes it run `21x21x20x(~20+20*~10+20+100*4*~2)=~9mln` elementary operations with rather high memory pressure.
Any ideas how to make it more moderate?

ilya-g commented:
Hi, @zhelenskiy. Thanks for the PR, this is a good start, but there are some problems with compilation and tests, especially in JS. It also has got new conflicts with master which are to be resolved.
I think I'm gonna take over this branch and try to fix these problems myself. Could you please help by answering my questions above?

zhelenskiy reviewed (commented):
I have answered the questions. Should I fix it myself or would you fix it yourself?

ilya-g commented on libraries/stdlib/test/ranges/RangeTest.kt:
```diff
@@ -5,11 +5,198 @@
+                    .takeWhile { if (s > 0) start <= it && it <= finish else start >= it && it >= finish }.toList()
+                try {
+                    collectionWithIterating(listRange, range, limitRange)
+                    collectionWithoutIterating(listRange, object : Collection<T> {
```
A lot of range and progression iteration cases are already covered in [`RangeIterationTest`](https://github.com/jetbrains/kotlin/blob/651d6c9d514e6ac620f1f18ff32728a6c050a573/libraries/stdlib/test/ranges/RangeIterationTest.kt). I don't think we should be duplicating iteration tests here considering that we didn't change the implementation of iterator.

zhelenskiy replied:
I have just looked there. It really repeats collectionWithIterating so it can be remobed but doesn't do the same with `without`. `Without` version checks size, contains and containsAll while `RangeIterationTestBase` doesn't. It also checks, that real iterator is not used. Furthermore, it is applied to big progressions which cannot be fully stored in memory.

I suppose that
```kotlin
if (expectedElements.isEmpty())
    assertTrue(sequence.none())
else
    assertEquals(expectedElements, sequence.toList())
```
should be replaced with
```kotlin
assertEquals(expectedElements.isEmpty(). sequence.none())
assertEquals(expectedElements, sequence.toList())
```

in [RangeIterationTest.kt](https://github.com/JetBrains/kotlin/blob/651d6c9d514e6ac620f1f18ff32728a6c050a573/libraries/stdlib/test/ranges/RangeIterationTest.kt#L77) as isEmpty is defined separately from iterator and progression implementation that always returns true would pass your test.

ilya-g replied:
Thanks for the suggestion, makes sense to apply it to `RangeIterationTestBase.doTest`

ilya-g replied:
Therefore, I'll remove `collectionWithIterating` tests and simplify the remaining ones.

zhelenskiy replied:
Great!

ilya-g commented:
I'll take the remaining issues myself. The current state of branch can be monitored here: https://github.com/JetBrains/kotlin/compare/rr/stdlib/range-collection

ilya-g commented:
The branch was ready to merge, but we decided to [hold it off](https://youtrack.jetbrains.com/issue/KT-36932#focus=Comments-27-5229742.0-0) until we figure how to provide a more smooth migration for the code affected by the change.
Since no action is longer needed in this PR, I'm going to close it now.
@zhelenskiy, thanks again for working on it.
